#creating a easier to read marker file from the GEMMA SNPs
library(dplyr)
library(stringr)
library(data.table)
library(parallel)

rm(list = ls())

#reading in the genotype data
setwd("/Users/jkhta/Desktop/nam_cam_fixing/27 - last_data/input/NAM_core_pops/gemma_FULL_NAM")
setwd("/home/jkta/projects/col_sha/nam_cam_markers_stated_creation")
marker_info <- fread("geno_FULL_NAM_lines_bimbam_format_no_header.tsv", 
                     sep = "\t")

pruned_markers <- readRDS("nam_rqtl_geno_prob_array_james_0.99_pruned_revised.rds")

#reading genotype data as lines because they're easier to change
cam_read <- file("geno_FULL_NAM_lines_bimbam_format_no_header.tsv", open = "r")
cam_raw <- readLines(cam_read)
cam_fix <- gsub("\t", " ", cam_raw)

#changes each row in the bimbam file into columns, and then binds all those columns together; 
#finally they change the columns into characters
test_fix_dup <- mclapply(cam_fix, function(x) as.data.frame(unlist(strsplit(x, split = " "))), mc.cores = 4)
test_do_call <- bind_cols(test_fix_dup)
test_do_call <- apply(test_do_call, 2, function(x) as.character(x))

colnames(test_do_call) <- test_do_call[1, ]
test_do_call_pruned <- test_do_call[2:nrow(test_do_call), ]
test_do_call_pruned <- subset(test_do_call_pruned, select = colnames(test_do_call_pruned) %in% names(pruned_markers))
  
test_states <- test_do_call[2:3, ]
test_states <- subset(test_states, select = colnames(test_states) %in% names(pruned_markers))

#reading in marker states data
nam_marker_states <- as.data.frame(fread("marker_states_for_Col0_and_ALT.csv", sep = ",", stringsAsFactors = FALSE))
nam_marker_states <- subset(nam_marker_states, marker %in% names(pruned_markers))

#converting the marker states into the correct snp states; this is so that tassel
#can use this information for its stepwise regression
control_state <- c("AA", "BB")

#converting the states into marker states
for (i in 1:ncol(test_states)) {
  marker_name <- colnames(test_states)[i]
  marker_info <- test_states[, i]
  marker_state <- subset(nam_marker_states, marker == marker_name)
  parent_state <- control_state[match(marker_info, marker_state[2:3])]
  parent_state <- parent_state[complete.cases(parent_state)]
  if (length(parent_state) > 0) {
    test_states[, i] <- parent_state
  } else {
    next
  }
}

nam_ril_marker_states <- test_do_call_pruned
#nam_ril_marker_states[1:2, ] <- test_states[1:2, ]

test_ril_marker_state <- nam_ril_marker_states[, 4]

ril_marker_replacer <- function(marker_vector) {
  marker_vector_df <- as.data.frame(marker_vector)
  for (i in 3:length(marker_vector)) {
    if (marker_vector[i] == "1") {
      marker_vector[i] <- NA
    } else if(marker_vector[i] == "0") {
      marker_vector[i] <- marker_vector[1]
    } else if(marker_vector[i] == "2") {
      marker_vector[i] <- marker_vector[2]
    }
  }
  return(marker_vector)
}

testing <- ril_marker_replacer(test_ril_marker_state)

#using the function to convert each of the marker column states into base pairs
nam_ril_marker_states_ref <- apply(nam_ril_marker_states, 2, function(x) ril_marker_replacer(x))
nam_ril_marker_states_ref[nam_ril_marker_states_ref == NA] <- "NA"

#reading in the genotype ril order to assign the genotypes to the appropriate rils
gemma_geno_ril_order <- as.data.frame(t(fread("geno_FULL_NAM_lines_bimbam_format_HEADER.txt", 
                                              sep = "\t", 
                                              header = FALSE)))
names(gemma_geno_ril_order) <- "rils"

#looking at the genotype only data
gemma_genos_only <- subset(gemma_geno_ril_order, grepl("RV", rils))
gemma_genos_only$rils <- gsub("X", "", gemma_genos_only$rils)

#getting the split between the rils in order to fill the rest wiht 0's
gemma_rils_split <- strsplit(gemma_genos_only$rils, split = "RV")
gemma_genos_only$pop_no <- sapply(gemma_rils_split, function(x) x[1])
gemma_genos_only$ril_no <- sapply(gemma_rils_split, function(x) x[2])
gemma_genos_only$ril_no <- str_pad(gemma_genos_only$ril_no, 3, side = "left", pad = 0)
gemma_genos_only$geno <- with(gemma_genos_only, paste(pop_no, "RV", ril_no, sep = ""))
gemma_genos_only$geno <- factor(gemma_genos_only$geno)
gemma_genos_only$pop <- paste(gemma_genos_only$pop_no, "RV", sep = "")
gemma_genos_only$geno_pop <- with(gemma_genos_only, paste(pop, geno, sep = "_"))


nam_genos <- gemma_genos_only$geno_pop

nam_ril_marker_geno_states_ref <- cbind(as.character(nam_genos), nam_ril_marker_states_ref[3:nrow(nam_ril_marker_states), ])
nam_ril_marker_geno_states_ref <- as.data.frame(nam_ril_marker_geno_states_ref)

#nam_ril_marker_geno_states_ref <- as.data.table(nam_ril_marker_geno_states_ref)
colnames(nam_ril_marker_geno_states_ref)[1] <- "geno"

#this part was generated by substituting the parental and the other state by AA and BB
fwrite(as.data.table(nam_ril_marker_geno_states_ref), 
       file = "nam_cam_markers_stated_pruned_subset.csv", 
       col.names = TRUE, 
       sep = ",", 
       na = "NA")

#this file was generated by keeping the SNP states the same in order to generate a VCF file with the right
#SNPs and genotype information for RTM-GWAS and TASSEL

# write.csv(nam_ril_marker_geno_states_ref,
#           file = "nam_cam_markers_stated.csv")
